Automattic/wp-calypso
# Checkout: Convert new credit card payment method to TypeScript

#### Changes proposed in this Pull Request

This PR converts the "Credit or debit card" (AKA "new credit card") payment method to TypeScript.

For the most part this should not change anything significant, but several small changes did have to be made to satisfy the TS compiler. Notably, the `@wordpress/data` store for the payment method had to be renamed from `'credit-card'` to `'wpcom-credit-card'` because the Jetpack system took the former name for their store when they duplicated this one, and it seemed easier to change this store rather than theirs.

Other small changes include:

- Stripe elements do not support a `disabled` prop (they use `options.disabled` instead which we already support).
- The submit button can get the `stripe` and `stripeConfiguration` data itself with the `useStripe` hook so we do not need to pass these to the payment method creator.
- We do not need to pass `items` or `total` to the payment processor; that's a relic from a very old version of this system.

#### Testing instructions

This adds some unit tests but you can also test manually as follows.

First test the Stripe version of the payment method:

- Add a product to your cart and visit checkout.
- Select "Credit or debit card" as your payment method.
- Fill in the fields with invalid data and verify that you see errors displayed inline below the invalid fields.
- Fill in the fields with valid data and submit the form.
- Verify that it works correctly.

Next test the EBANX version of the payment method:

  - Set your currency to `BRL`.
  - Add a product to your cart and visit checkout.
  - In the contact information step, choose "Brazil" for the country and `61919-230` for the postal code.
  - In the final step, choose "Credit or debit card".
  - Verify that the form asks for all the Ebanx fields in addition to the credit card fields (Taxpayer Identification Number, Phone, etc.)
  - Use an [ebanx test card](https://developer.ebanx.com/docs/resources/testCreditCards/) and [test customer info](https://developer.ebanx.com/docs/resources/testCustomerData/).
  - Verify that the payment succeeds and that you are redirected to the thank-you page.
